### PR TITLE
Add publicSubnetIds and privateSubnetIds cluster opts. Update tests to use new awsx.ec2.Vpc API and new subnet opts

### PR DESCRIPTION
**Pull request by metral**
### Proposed changes

**\*It helps to review this PR by going through each commit separately.**\*

1. **feat(subnetIds): add new opts: publicSubnetIds and privateSubnetIds**

    This change intends to tackle 2 cases:
    - Simplifying cluster subnet defintion by explicitly accepting both
    public and private subnets, and
    - To properly build the cluster dependency graph w.r.t. the subnets used by the
    cluster. This should produce expected results for nodegroup subnet definition
    between initial updates and proceeding empty preview steps.

    Cluster worker subnet definition benefits from having individual `publicSubnetIds`
    and `privateSubnetIds` options, as they are:
    - Explicit in defintion of which subnets are being used for a given subnet class.
    - Inline with user practice, and
    - `computeWorkerSubnets()` in the NodeGroup class is an approximation
    of the which subnets belong to which class, and it can occassionaly experience
    non-deterministic issues.
1. **fix(tests): update awsx.Network -> awsx.ec2.Vpc & use specific subnetIds**
    - Update tests that use VPCs created with `awsx.Network` to the newer
    `awsx.ec2.Vpc` API.
    - Remove any `awsx.ec2.Vpc` settings already set by defaults.
    - Update EKS cluster definitions that used the `eks.Cluster.subnetIds` to
    now use the new subnet options: `eks.Cluster.publicSubnetIds` and `eks.Cluster.privateSubnetIds`.
1. **test(awsx-network-and-subnetIds): add test of older VPC & network setup**
    Add new test which creates EKS clusters using the previous `awsx.Network` API
    and related cluster network setups.

    This test creates two types of clusters that use the previous network API setup:
    - One cluster uses a VPC from the `awsx.Network` API, and cluster `subnetIds`.
    - One cluster uses a VPC from the `awsx.Network` API, cluster
    `subnetIds`, and a node group with the `nodeSubnetIds` override option.
### Related issues

Closes https://github.com/pulumi/pulumi-eks/issues/199
Closes https://github.com/pulumi/pulumi-eks/issues/179

### CODE REVIEW

File: nodejs/eks/cluster.ts [hunk 1]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
+     * account's default VPC subnets.
+     *
+     * If the list of subnets includes both public and private subnets, the worker
+     * nodes will only be attached to the private subnets, and the public
+     * subnets will be used for internet-facing load balancers.
+     *
+     * See for more details: https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html.
+     *
+     * Note, the use of `subnetIds`, and the combination of `publicSubnetIds`
+     * and/or `privateSubnetIds` is mutually exclusive. The former is the
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: Might reword this to something like "This option will be deprecated in a future release." The former/latter wording is a little tricky to parse.

File: nodejs/eks/cluster.ts [hunk 2]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
+     * is newer functionality.
      */
     subnetIds?: pulumi.Input<pulumi.Input<string>[]>;
 
+    /**
+     * The set of public subnets to use on the cluster. These subnets
+     * are used as the default public subnets for the worker node groups, and
+     * are automatically tagged by EKS for Kubernetes purposes.
+     *
+     * Note, `vpcId` must be set. If not, then the cluster will use the AWS
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: I'd reword to
> If `vpcId` is not set, the cluster will use the AWS account's default VPC subnets.

File: nodejs/eks/cluster.ts [hunk 3]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
 
     /**
-     * The subnets to attach to the EKS cluster. If either vpcId or subnetIds is unset, the cluster will use the
-     * default VPC's subnets. If the list of subnets includes both public and private subnets, the Kubernetes API
-     * server and the worker nodes will only be attached to the private subnets. See
-     * https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html for more details.
+     * The set of subnets to use on the EKS cluster.
+     * These subnets are automatically tagged by EKS for Kubernetes purposes.
+     *
+     * Note, `vpcId` must be set. If not, then the cluster will use the AWS
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: I'd reword to
> If `vpcId` is not set, the cluster will use the AWS account's default VPC subnets.

File: nodejs/eks/cluster.ts [hunk 4]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
+     * account's default VPC subnets.
+     *
+     * Network architecture options:
+     *  - Private-only: Only set `privateSubnetIds`.
+     *    - Default workers to run in a private subnet and Kubernetes cannot create internet-facing load
+     *  balancers for your pods.
+     *  - Public-only: Only set `publicSubnetIds`.
+     *    - Default workers to run in a public subnet, including your worker nodes.
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: "All nodes run in a public subnet, including master nodes"? Not sure what is correct, but something seems off here.

File: nodejs/eks/cluster.ts [hunk 5]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
+     *  - Mixed (recommended): Set both `privateSubnetIds` and `publicSubnetIds`.
+     *    - Default all worker nodes to run in private subnets, and use the public subnets
+     *  for internet-facing load balancers.
+     *
+     * See for more details: https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html.
+     *
+     * Note, the use of `subnetIds`, and the combination of `publicSubnetIds`
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: Recommend rewording as suggested above.

File: nodejs/eks/cluster.ts [hunk 6]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
+     * is newer functionality.
+     */
+    publicSubnetIds?: pulumi.Input<pulumi.Input<string>[]>;
+
+    /*
+     * The set of private subnets to use on the cluster. These subnets
+     * are used as the default private subnets for the worker node groups, and
+     * are automatically tagged by EKS for Kubernetes purposes.
+     *
+     * Note, `vpcId` must be set. If not, then the cluster will use the AWS
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: I'd reword to
> If `vpcId` is not set, the cluster will use the AWS account's default VPC subnets.

File: nodejs/eks/cluster.ts [hunk 7]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
+     * account's default VPC subnets.
+     *
+     * Network architecture options:
+     *  - Private-only: Only set `privateSubnetIds`.
+     *    - Default workers to run in a private subnet and Kubernetes cannot create internet-facing load
+     *  balancers for your pods.
+     *  - Public-only: Only set `publicSubnetIds`.
+     *    - Default workers to run in a public subnet, including your worker nodes.
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: Same comment as above.

File: nodejs/eks/cluster.ts [hunk 8]
```diff
@@ -484,13 +512,82 @@ export interface ClusterOptions {
...
+     *  - Mixed (recommended): Set both `privateSubnetIds` and `publicSubnetIds`.
+     *    - Default all worker nodes to run in private subnets, and use the public subnets
+     *  for internet-facing load balancers.
+     *
+     * See for more details: https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html.
+     *
+     * Note, the use of `subnetIds`, and the combination of `publicSubnetIds`
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]: Same comment as above.

File: nodejs/eks/nodegroup.ts [hunk 9]
```diff
@@ -44,9 +44,16 @@ export interface Taint {
 export interface NodeGroupBaseOptions {
 
     /**
-     * The IDs of the explicit node subnets to attach to the worker node group.
+     * The explicit subnet IDs to use for the worker node group.
      *
-     * This option overrides clusterSubnetIds option.
+     * The IDs used must be a subset, or the whole set of IDs passed to the
```
Comment (lblackstone) [inline nodejs/eks/nodegroup.ts]: This section is confusing. It looks like this overrides default behavior, but the explanation seems overly complex. I'd focus on what it overrides.

File: nodejs/eks/nodegroup.ts [hunk 10]
```diff
@@ -448,7 +455,26 @@ ${customUserData}
...
+        core.subnetIds,
+    ]).apply(([overrideIds, privateIds, publicIds, subnetIds]) => {
+        if (overrideIds) { // Use the specified override subnetIds.
+            return pulumi.output(overrideIds);
+        } else if (privateIds) { // Use the specified private subnetIds.
+            return pulumi.output(privateIds);
+        } else if (publicIds) { // Use the specified public subnetIds.
+            return pulumi.output(publicIds);
+        }
+        // Use subnetIds from the cluster. Compute / auto-discover the private worker subnetIds from this set.
```
Comment (lblackstone) [inline nodejs/eks/nodegroup.ts]: Probably want to mention that this is the legacy default behavior

File: nodejs/eks/examples/tests/awsx-network-and-subnetIds/iam.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+import * as pulumi from "@pulumi/pulumi";
+
+const managedPolicyArns: string[] = [
+    "arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy",
+    "arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy",
+    "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly",
+];
+
+// Creates a role and attaches the EKS worker node IAM managed policies
+export function createRole(name: string): aws.iam.Role {
```
Comment (lblackstone) [inline nodejs/eks/examples/tests/awsx-network-and-subnetIds/iam.ts]: Does it make sense to include this as part of the EKS lib rather than defining in the test?

File: nodejs/eks/cluster.ts [hunk 12]
```diff
@@ -223,14 +225,38 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
...
     }
 
-    // If no VPC was specified, use the default VPC.
+    if (args.subnetIds && (args.publicSubnetIds || args.privateSubnetIds)) {
+        throw new Error("subnetIds, and the use of publicSubnetIds and/or privateSubnetIds are mutually exclusive. Choose a single approach.");
+    }
+
+    // Configure default networking architecture.
     let vpcId: pulumi.Input<string> = args.vpcId!;
     let subnetIds: pulumi.Input<pulumi.Input<string>[]> = args.subnetIds!;
```
Comment (lukehoban) [inline nodejs/eks/cluster.ts]: What is `subnetIds` used for now?

File: nodejs/eks/cluster.ts [hunk 13]
```diff
@@ -223,14 +225,38 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
...
         vpcId = vpc.then(v => v.id);
         subnetIds = vpc.then(v => aws.ec2.getSubnetIds({ vpcId: v.id }, invokeOpts)).then(subnets => subnets.ids);
+        clusterSubnetIds = subnetIds;
+    }
+
+    // Form the subnetIds to use on the cluster from either:
+    //  - subnetIds
+    //  - A combination of privateSubnetIds and/or publicSubnetIds.
+    if (args.subnetIds) {
+        clusterSubnetIds = subnetIds;
```
Comment (lukehoban) [inline nodejs/eks/cluster.ts]: It looks wrong that this uses `args.subnetIds` for the check but `subnetIds` in the assignment.

I honestly can't quite follow how this logic intersects with the population of `subnetIds` in the default VPC case above.

File: nodejs/eks/cluster.ts (hunk 1 above)
Comment (lukehoban) [reply nodejs/eks/cluster.ts]: I'm not sure we need to say it will be deprecated, or that there is "old"/"new" functionality.  We can just be fact based about what the legal combinations are and what they do.  If we want to say "passing publicSubnetIds and privateSubnetsIds is encouraged" that seems fine.

File: nodejs/eks/examples/tests/awsx-network-and-subnetIds/iam.ts (hunk 11 above)
Comment (metral) [reply nodejs/eks/examples/tests/awsx-network-and-subnetIds/iam.ts]: We had the same question be asked by a user. Here's the thread:

- https://github.com/pulumi/pulumi-eks/pull/98#issuecomment-479978253
- https://github.com/pulumi/pulumi-eks/pull/98#issuecomment-480452516

In short:

> The approach here gives users who want to control the set of roles being used complete control over those roles and the associated instance profiles. It looks like the most flexible way to enable this while ensuring that all the related requests to this are now 

It's been mentioned to possibly embed this in `pulumi/awsx` in the past, but it never materialized. We've since copied around this `iam.ts` for the examples & tests that need them in `pulumi/eks` for node groups.

File: nodejs/eks/cluster.ts (hunk 12 above)
Comment (metral) [reply nodejs/eks/cluster.ts]: The primary use of `subnetIds` is now only for:

1) backwards compatibility for users still on `subnetIds`:
2) and for use in the default VPC case when no `nodeSubnetIds` overrides are set, or if `publicSubnetIds` and/or `privateSubnetIds` is not set per the PR changes:

See:

https://github.com/pulumi/pulumi-eks/blob/18b55dd9dc6e03595d8eec3057f1eab1ae7892a5/nodejs/eks/cluster.ts#L240-L246

and

https://github.com/pulumi/pulumi-eks/blob/18b55dd9dc6e03595d8eec3057f1eab1ae7892a5/nodejs/eks/nodegroup.ts#L460-L474

File: nodejs/eks/cluster.ts (hunk 13 above)
Comment (metral) [reply nodejs/eks/cluster.ts]: > It looks wrong that this uses args.subnetIds for the check but subnetIds in the assignment.

I've updated the checks, PTAL.

> I honestly can't quite follow how this logic intersects with the population of subnetIds in the default VPC case above.

See comment [above](https://github.com/pulumi/pulumi-eks/pull/239#discussion_r320446888).

**Comment by metral** [comment]
Feedback has been addressed. PTAL @lukehoban @lblackstone

File: nodejs/eks/cluster.ts [hunk 14]
```diff
@@ -223,14 +225,38 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
...
+    const privateSubnetIds: pulumi.Input<pulumi.Input<string>[]> = args.privateSubnetIds!;
+    let clusterSubnetIds: pulumi.Input<pulumi.Input<string>[]> = [];
+
+    // If no VPC is set, use the default VPC's subnets.
+    if (!args.vpcId) {
         const invokeOpts = { parent: parent };
         const vpc = aws.ec2.getVpc({ default: true }, invokeOpts);
         vpcId = vpc.then(v => v.id);
         subnetIds = vpc.then(v => aws.ec2.getSubnetIds({ vpcId: v.id }, invokeOpts)).then(subnets => subnets.ids);
+        clusterSubnetIds = subnetIds;
```
Comment (lukehoban) [inline nodejs/eks/cluster.ts]: There still seems to be something unnecessarily convoluted going on here. 

At the very least - this line is unnecessary. 

But it also still seems like in fact the entire use of subnetIds is unnecessary. Like if you just removed it and then inlined all the places that use it - everything would be the same but simpler. 

In the note above you point to code in a different file that uses core.subnetIds, but I’m talking about the subnetIds local variable here which is unrelated.

File: nodejs/eks/cluster.ts [hunk 15]
```diff
@@ -484,13 +512,75 @@ export interface ClusterOptions {
...
+     * See for more details: https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html.
+     *
+     * Note: The use of `subnetIds`, along with `publicSubnetIds`
+     * and/or `privateSubnetIds` is mutually exclusive. The use of
+     * `publicSubnetIds` and `privateSubnetIds` is encouraged.
+     */
+    publicSubnetIds?: pulumi.Input<pulumi.Input<string>[]>;
+
+    /*
+     * The set of public subnets to use for the worker node groups on the EKS cluster.
```
Comment (lblackstone) [inline nodejs/eks/cluster.ts]:
```suggestion
     * The set of private subnets to use for the worker node groups on the EKS cluster.
```

File: nodejs/eks/examples/tests/awsx-network-and-subnetIds/package.json [hunk 16]
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "test-awsx-network-and-subnetIds",
+    "devDependencies": {
+        "typescript": "^3.0.0",
+        "@types/node": "latest"
+    },
+    "dependencies": {
+        "@pulumi/pulumi": "^1.0.0-beta",
```
Comment (lblackstone) [inline nodejs/eks/examples/tests/awsx-network-and-subnetIds/package.json]: Should be able to use 1.0.0 versions now

File: nodejs/eks/cluster.ts (hunk 14 above)
Comment (metral) [reply nodejs/eks/cluster.ts]: I've removed the use of the local `subnetIds` variable, and inlined it. `subnetIds` only gets used in the following places:

- https://github.com/pulumi/pulumi-eks/blob/a869af60fc1a61049add65d622dc3202abfedd79/nodejs/eks/cluster.ts#L251
- https://github.com/pulumi/pulumi-eks/blob/a869af60fc1a61049add65d622dc3202abfedd79/nodejs/eks/cluster.ts#L474

File: nodejs/eks/examples/tests/awsx-network-and-subnetIds/package.json (hunk 16 above)
Comment (metral) [reply nodejs/eks/examples/tests/awsx-network-and-subnetIds/package.json]: Done

**Comment by metral** [comment]
Feedback has been addressed. PTAL @lukehoban @lblackstone

File: nodejs/eks/cluster.ts [hunk 17]
```diff
@@ -84,6 +84,8 @@ export interface CoreData {
     cluster: aws.eks.Cluster;
     vpcId: pulumi.Output<string>;
     subnetIds: pulumi.Output<string[]>;
+    publicSubnetIds: pulumi.Output<string[]>;
+    privateSubnetIds: pulumi.Output<string[]>;
```
Comment (pgavlin) [inline nodejs/eks/cluster.ts]: Seems like both `publicSubnetIds` and `privateSubnetIds` should be optional:

```typescript
    publicSubnetIds?: pulumi.Output<string[]>;
    privateSubnetIds?: pulumi.Output<string[]>;
```

File: nodejs/eks/cluster.ts [hunk 18]
```diff
@@ -223,14 +225,37 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
...
+        throw new Error("subnetIds, and the use of publicSubnetIds and/or privateSubnetIds are mutually exclusive. Choose a single approach.");
+    }
+
+    // Configure default networking architecture.
     let vpcId: pulumi.Input<string> = args.vpcId!;
-    let subnetIds: pulumi.Input<pulumi.Input<string>[]> = args.subnetIds!;
-    if (args.vpcId === undefined) {
+    const subnetIds: pulumi.Input<pulumi.Input<string>[]> = args.subnetIds!;
+    const publicSubnetIds: pulumi.Input<pulumi.Input<string>[]> = args.publicSubnetIds!;
+    const privateSubnetIds: pulumi.Input<pulumi.Input<string>[]> = args.privateSubnetIds!;
```
Comment (pgavlin) [inline nodejs/eks/cluster.ts]: I don't think we need these locals--they're only ever set to the input args. Instead, the various refs to these can use the arguments directly.

File: nodejs/eks/cluster.ts [hunk 19]
```diff
@@ -447,6 +472,8 @@ export function createCore(name: string, args: ClusterOptions, parent: pulumi.Co
     return {
         vpcId: pulumi.output(vpcId),
         subnetIds: pulumi.output(subnetIds),
+        publicSubnetIds: pulumi.output(publicSubnetIds),
+        privateSubnetIds: pulumi.output(privateSubnetIds),
```
Comment (pgavlin) [inline nodejs/eks/cluster.ts]: Feels like these should be `args.publicSubnetIds ? pulumi.output(args.publicSubnetIds) : undefined` and args.privateSubnetIds ? pulumi.output(args.privateSubnetIds) : undefined`, respectively.

File: nodejs/eks/nodegroup.ts [hunk 20]
```diff
@@ -448,7 +450,26 @@ ${customUserData}
         userData: userdata,
     }, { parent: parent, ignoreChanges: ignoreChanges });
 
-    const workerSubnetIds = args.nodeSubnetIds ? pulumi.output(args.nodeSubnetIds) : pulumi.output(core.subnetIds).apply(ids => computeWorkerSubnets(parent, ids));
+    // Compute the worker node group subnets to use from the various approaches.
+    let workerSubnetIds: pulumi.Output<string[]> = pulumi.output([]);
+    workerSubnetIds = pulumi.all([
```
Comment (pgavlin) [inline nodejs/eks/nodegroup.ts]: This doesn't need to be inside of an `all/apply` if we make `core.privateSubnetIds` and `core.publicSubnetIds` optional. With that change, you can just write this:

```typescript
let workerSubnetIds: pulumi.Output<string[]>;
if (args.nodeSubnetIds !== undefined) {
    workerSubnetIds = pulumi.output(args.nodeSubnetIds);
} else if (core.privateSubnetIds !== undefined) {
    workerSubnetIds = core.privateSubnetIds;
} else if (core.publicSubnetIds !== undefined) {
    workerSubnetIds = core.publicSubnetIds;
} else {
    workerSubnetIds = pulumi.output(subnetIds).apply(ids => computeWorkerSubnets(parent, ids));
}
```

File: nodejs/eks/cluster.ts (hunk 18 above)
Comment (metral) [reply nodejs/eks/cluster.ts]: Done

File: nodejs/eks/cluster.ts (hunk 19 above)
Comment (metral) [reply nodejs/eks/cluster.ts]: Done

File: nodejs/eks/nodegroup.ts (hunk 20 above)
Comment (metral) [reply nodejs/eks/nodegroup.ts]: Done

**Comment by metral** [comment]
Feedback has been addressed. PTAL @pgavlin

File: nodejs/eks/cluster.ts (hunk 17 above)
Comment (metral) [reply nodejs/eks/cluster.ts]: Done
